src/adm.rs

```diff
@@ -1,39 +1,269 @@
+                }
+                match self.check_impression(impression_filter, &mut result, tags) {
+                    Ok(_) => {}
+                    Err(e) => {
+                        self.report(&e, tags);
+                        return None;
+                    }
+                }
+                result.position = if filter.position.is_none() {
+                    default.position
+                } else {
+                    filter.position
+                };
+                Some(result)
+            }
+            None => {
+                self.report(
+                    &HandlerErrorKind::UnexpectedAdvertizer(tile.name).into(),
+                    tags,
+                );
+                None
+            }
+        }
+    }
+}
+
+/// Construct the AdmFilter from the provided settings.
+/// This uses a JSON construct of settings, e.g.
+/// ```javascript
+/// /* for the Example Co advertiser... */
+/// {"Example": {
+///     /* The allowed hosts for URLs */
+///     "advertiser_hosts": ["www.example.org", "example.org"],
+///     /* Valid tile positions for this advertiser (empty for "all") */
+///     "positions": [1, 2],
+///     /* Valid target regions for this advertiser
+///        (use "en-US" for "all in english speaking United States") */
+///     "include_regions": ["en-US/TX", "en-US/CA"],
+///     /* Allowed hosts for impression URLs.
+///        Empty means to use the impression URLs in "DEFAULT" */
+///     "impression_hosts: [],
+///     },
+///     ...,
+///  "DEFAULT": {
+///    /* The default impression URL host to check for. */
+///    "impression_hosts": ["example.net"]
+///     }
+/// }
+/// ```
+///
+impl From<&Settings> for HandlerResult<AdmFilter> {
+    fn from(settings: &Settings) -> Self {
+        let mut filter_map: BTreeMap<String, AdmAdvertiserFilterSettings> = BTreeMap::new();
+        for (adv, setting) in settings.adm_settings.clone() {
+            dbg!("Processing records for {:?}", &adv);
+            // map the settings to the URL we're going to be checking
+            filter_map.insert(adv.to_lowercase(), setting);
+        }
+        Ok(AdmFilter {
+            filter_set: filter_map,
+        })
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
 pub struct AdmTile {
     pub id: u64,
     pub name: String,
     pub advertiser_url: String,
     pub click_url: String,
     pub image_url: String,
     pub impression_url: String,
+    pub position: Option<u8>,
 }
 
 pub async fn get_tiles(
     reqwest_client: &reqwest::Client,
     adm_endpoint_url: &str,
-    fake_ip: &str,
+    location: &LocationResult,
     stripped_ua: &str,
     placement: &str,
+    state: &ServerState,
+    tags: &mut Tags,
 ) -> Result<AdmTileResponse, HandlerError> {
     // XXX: Assumes adm_endpoint_url includes
     // ?partner=<mozilla_partner_name>&sub1=<mozilla_tag_id> (probably should
     // validate this on startup)
+    // let user_loc =
+    let settings = state.settings.clone();
```

Comment:
nit: is it possible to only clone the relevant parts (`partner_id` and `sub1`) here?

Reply:
Hrm, wondering if we should only query adm for 2 as well, or do we expect to have to toss out tiles a lot?

Reply:
I think we'll need to query adm for a result count larger than our `max_size` due to our filtering. They're not serving very many tiles so your guess is as good as mine on how many to ask for -- a few, maybe 10.

Reply:
![settingsman](https://user-images.githubusercontent.com/35755/116758253-8087ce00-a9c4-11eb-89d5-7cba41a73e11.png)